atlante/atlante.go

```diff
@@ -21,31 +23,62 @@ import (
 )
 
 type ImgStruct struct {
-	Filename string
-	Width    int
-	Height   int
+	filename string
+	lck      sync.Mutex
+	// Have we generated the file?
+	// This will allow us to generate the file only when requested
+	generated bool
+	image     *image.Image
+	filestore filestore.FileWriter
+	Width     int
+	Height    int
 }
 
-func (img ImgStruct) ImageTagBase64() string {
-
-	return fmt.Sprintf(
-		`<image width="%d" height="%d" xlink:href="data:%s;base64,%s" />`,
-		img.Width,
-		img.Height,
-		img.MimeType(),
-		img.Base64Image(),
-	)
+// Filename returns the file name of the image
+func (img ImgStruct) Filename() (string, error) {
+	if img.generated {
+		return img.filename, nil
+	}
+	// We need to generate the file and then return the filename
+	return img.generateImage()
 }
 
-func (img ImgStruct) MimeType() string { return "image/png" }
-
-func (img ImgStruct) Base64Image() string {
-	filebytes, err := ioutil.ReadFile(img.Filename)
+// generateIamge is use to create the image into the filestore
+func (img *ImgStruct) generateImage() (fn string, err error) {
+	// Make sure generateImage is only called one at a time.
```

Comment:
is this a hack for some other problem? I find the logic a bit complicated to follow.